stan/math/opencl/rev/arena_matrix_cl.hpp

```diff
@@ -0,0 +1,52 @@
+#ifndef STAN_MATH_OPENCL_REV_ARENA_MATRIX_CL_HPP
+#define STAN_MATH_OPENCL_REV_ARENA_MATRIX_CL_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/rev/core/chainable_alloc.hpp>
+#include <stan/math/opencl/kernel_generator/is_kernel_expression.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/**
+ * A variant of `matrix_cl` that schedules its destructor to be called, so it
+ * can be used on the AD stack.
+ */
+template <typename T>
+class arena_matrix_cl : public chainable_alloc, public matrix_cl<T> {
```

Comment:
Is there a reason this is it's own class and not a specialization of `arena_matrix_cl`? My though process was that if we can have reverse mode functions that work with `arena_matrix` then they could also accept `matrix_cl`. Is there a reason we would want to have explicit versions that need to call this version of `arena_matrix_cl`?

Reply:
> Is there a reason this is it's own class and not a specialization of arena_matrix_cl?

I don't understand Are you asking why `arena_matrix_cl` is not specialization of `arena_matrix_cl`?

> we can have reverse mode functions that work with arena_matrix then they could also accept matrix_cl

It might be possible to do that for some functions, but `arena_matrix_cl` and `arena_matrix` do not have exactly the same interface.

> Is there a reason we would want to have explicit versions that need to call this version of arena_matrix_cl?

What do you mean by explicit versions and this version of `arena_matrix_cl`? There is no other version of `arena_matrix_cl`.

Reply:
> I don't understand Are you asking why arena_matrix_cl is not specialization of arena_matrix_cl?

Ack sorry massive typo I meant `arena_matrix`. I also misread earlier and see now this is being used in `arena_type_impl`

> It might be possible to do that for some functions, but arena_matrix_cl and arena_matrix do not have exactly the same interface.

I'll resolve this comment for now as that's what I want to look at but is separate.